[GSoC 2021] Socket Library

king-11 opened the pull request:
This PR adds an implementation of the Socket library to modules/packages.

This first version has some caveats:

 * requires libevent 2.1 or newer
 * requires CHPL_TASKS=qthreads right now (fifo doesn't work yet)
 * requires the C backend right now (CHPL_TARGET_COMPILER=llvm doesn't work yet)
 * requires libevent be available in /usr/include/event2
 
### Design 
#17899 

### Future Work
  - DNS Resolution using libevent
  - IO Plugin using libevent for non-blocking IO
  - sendAll method for UDP and TCP on lines of python's [sendAll](https://docs.python.org/3/library/socket.html#socket.socket.send)
  - Make work with LLVM and FIFO
  - Separate Module for Libevent

ankingcodes commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,481 @@
+  enum IPFamily {
+    IPv4 = 2,
+    IPv6 = 10,
+    IPUnspec = 0
```
I assume IPUnspec is an enum named by you, if so, I feel "IPNoSpec" would be better.

king-11 replied:
It's a POSIX translation of `AF_UNSPEC` as we can't have `c_int` in enum I used their values directly

ankingcodes reviewed (commented):
Did a review pass mostly over TCP part. Looks good. Made some inline comments. 
@king-11 Looking forward for tests. For documentation, I believe you'll be writing doc-strings as comments over the functions. Also, please resolve merge conflicts.

king-11 commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,481 @@
+  }
+
+  pragma "no doc"
+  proc ipAddr.init(in address:sys_sockaddr_t) {
+    this._addressStorage = new sys_sockaddr_t();
+    try! {
+      _addressStorage.set(address.numericHost().c_str(), address.port(), address.family);
+    }
+  }
```
The use of this would for inside Module use mostly as to copy an ipAddr or create one we essentially need to copy the `_addressStorage` inside it which is returned from system calls

king-11 commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,481 @@
+    var _addressStorage:sys_sockaddr_t;
+  }
+
+  proc ipAddr.init(inout host:string = "localhost", port:uint(16) = 8000, family:IPFamily = IPFamily.IPv4) {
+    if(host == "localhost"){
+      host = if family == IPFamily.IPv4 then "127.0.0.1" else "::1";
+    }
+
+    _addressStorage = new sys_sockaddr_t();
+
+    try! {
+      _addressStorage.set(host.c_str(), port, family:c_int);
+    }
+  }
```
sure

rapiz1 commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1233 @@
+}
+
+/* compare ipAddr */
+inline operator !=(in lhs: ipAddr,in rhs: ipAddr) throws {
```
Why does this throw?

king-11 replied:
it's possible that the user might have not configured the ipAddr properly so the getters for `family`, `host` and `port` are throwing that's why the comparison operator is also throwing

king-11 replied:
I can add in a `try!` here because the `ipAddr` will set properly but my thoughts were let's say a user does something like
```chapel
var addr1:ipAddr, addr2:ipAddr;
writeln(addr1 != addr2);
```
Basically, it will throw or halt (will it ?) if we use `try!` when the user uses an uninitialized `ipAddr` record.

mppf replied:
We can create a `ipAddr.init()` that default initializes the record with 0.0.0.0 e.g. so that these do not have to throw/halt in that case. IMO it will be better if the initializer guarantees that the ipAddr is valid enough for the getters not to throw, and then neither comparison nor comparison will throw.

mppf commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1282 @@
+  :rtype: `ipAddr`
+  :throws SystemError: Upon incompatible `host`, `port` or `family`
+*/
+proc type ipAddr.create(host: string = "127.0.0.1", port: uint(16) = 8000, family: IPFamily = IPFamily.IPv4): ipAddr throws {
```
I was expecting one to write `new ipAddr("127.0.0.1", 8111, IPFamily.IPv4)` etc. See https://github.com/chapel-lang/chapel/pull/17899#issuecomment-923007197

king-11 replied:
Added the comment in design doc

mppf replied:
Can you add a comment here as well please? (It could be inside of the function - it is just for maintainers to keep track)

mppf commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1282 @@
+  var tempfd:c_int;
+  var err:syserr = ENOERR;
+  on this.home {
+    err = qio_get_fd(this._file_internal, tempfd);
```
You need to create a local temporary fd here so that the last argument to qio_get_fd is local.

mppf commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1282 @@
+  :rtype: `ipAddr`
+*/
+proc tcpConn.addr throws {
+  return getpeername(this.socketFd);
```
If the `tcpConn` is on a different locale, this function will fail, because the file descriptor will not be know to the system on the other locale. I think you need to include an `on` statement here.

king-11 replied:
compiler won't allow me to have return statement inside of an `on` statement how should I handle that

mppf replied:
You have to create a temporary variable, e.g.

``` chpl
proc myproc() {
  var tmp: int;
  on somewhere {
    tmp = computeSomething();
  }
  return tmp;
}
```

mppf commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1282 @@
+
+inline operator !=(in lhs: tcpConn,in rhs: tcpConn) {
+  return lhs.socketFd != rhs.socketFd;
+}
```
How about `==` as well?

king-11 replied:
sounds right

mppf commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1282 @@
+private extern proc evutil_make_socket_nonblocking(fd: c_int):c_int;
+private extern proc libevent_global_shutdown();
+
+extern type pthread_t = c_ulong;
```
This one seems odd to me. As far as I can tell, the POSIX standard doesn't say anything about what type `pthread_t` should be. It could be a struct. What happens if it is just `extern type pthread_t` ?

king-11 replied:
I will make the change for it

mppf commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1282 @@
+  if x != 0 {
+    throw new Error("event loop wasn't initialized");
+  }
+  pthread_exit(c_ptrTo(ret_val):c_void_ptr);
```
I would expect that simply returning is sufficient here, rather than calling `pthread_exit`. Do you know of a reason that is not the case?

Note though that if you want to return an integer, you have to copy it in to a `c_void_ptr`, since you can't return a pointer to a local variable without causing memory errors. So I'd recommend writing it something like this:

``` chapel
var retInt : c_intptr = 0;
var retAsPtr : c_void_ptr;
c_memcpy(retAsPtr, retInt, sizeof(retAsPtr));
return retAsPtr;
```

You can use a similar `c_memcpy` approach to get the return value out, but it appears today that you do not use it. If you know you don't need a return value from the `dispatchLoop` here, let's just have it return `nil` instead of doing `pthread_exit` and remove the `ret_val` variable entirely.

king-11 replied:
Will go with `nil` approach

mppf commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1282 @@
+  event_base_loopbreak(event_loop_base);
+  pthread_join(event_loop_thread, nil:c_ptr(c_void_ptr));
+  event_base_free(event_loop_base);
+  libevent_global_shutdown();
```
I would imagine that this would cause problems for other modules also wishing to use libevent. It seems like we would need a helper module for the libevent parts and that can manage calling `evthread_use_pthreads();` and `libevent_global_shutdown();`.

king-11 replied:
Marking as future work because we can then integrate libevent into runtime as well while tranferring it to a separate module.

mppf commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1282 @@
+*/
+module Socket {
+
+// TODO: replace with mason config
```
Somewhere, we need a description of what the implementation is doing with libevent at a high level. It could be here in a comment (that is not part of the docs) or it could be in the PR message itself.

king-11 replied:
I am not sure what you mean by implementation

mppf replied:
Like, can you make a comment that is a sort of road map / outline for this `Socket.chpl` file? I'm not so worried about outlining the different sections of implementation. I'd like it to describe how the Socket module is using libevent, at a high level. (The low level description can be the source code, but a high level view can be hard for someone reading the source code to get).

mppf commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1282 @@
+  var err_out:err_t = 0;
+  // try accept
+  err_out = sys_accept(socketFd, client_addr, fdOut);
+  // if error is not about blocking throw error
```
add a comma --

`if error is not about blocking, throw error`

mppf commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1282 @@
+      }
+      throw SystemError.fromSyserr(err_out, "accept() failed");
+    }
+    return this.accept();
```
Let's write this with a loop so that the structure is easier to see & so the stack memory usage does not grow with more attempts

mppf commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1282 @@
+  :return: bound address
+  :rtype: `ipAddr`
+*/
+proc tcpListener.addr throws {
```
It seems really weird that this can throw. Can you describe the situation a user would write that would cause it to throw? If it is not possible for a user to run in to it, we don't need this to throw.

king-11 replied:
my reasons for using this are the same as https://github.com/chapel-lang/chapel/pull/17960#discussion_r716154610 when user will use an uninitialized `tcpListener`

mppf replied:
Right. So we can make a `tcpListener.init()` that creates an acceptable default and ensure that the necessary properties are there for any other `init` for tcpListener.

mppf commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1282 @@
+
+inline operator !=(in lhs: tcpListener,in rhs: tcpListener) {
+  return lhs.socketFd != rhs.socketFd;
+}
```
How about `==` ? Also, it seems odd that this uses `in` intent. I think the default intent would be more appropriate here (and for a record such as tcpListener it is `const ref`).

king-11 replied:
I was using the same operator as I did in `ipAddr` so will change it over here, but is `in` required for `ipAddr` too or not?

mppf replied:
I don't think it's required in either case and I think the default is better in both of these cases.

mppf commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1282 @@
+/*
+  Default `backlog` value used in :proc:`listen`
+*/
+var BACKLOG_DEFAULT:uint(16);
```
I would expect this to be `const` rather than `var`.

Also, see https://github.com/chapel-lang/chapel/blob/main/doc/rst/developer/bestPractices/StandardModuleStyle.rst#other-identifiers -- I would expect this variable to be named `backlogDefault` -- we don't use the underscore naming style in standard modules.

mppf commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1282 @@
+pragma "no doc"
+extern const SOMAXCONN: int;
+/*
+  Default `backlog` value used in :proc:`listen`
```
It seems that the documentation here should say something about how we are computing this default value, even if it does so in a vague way.

mppf commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1282 @@
+/*
+  This overload of `connect` not only returns a :type:`tcpConn`
+  but also does DNS resolution for the provided `host`.
+  The `timeout` is tired for all resolved addresses and the first
```
Not sure what you mean by `tired` - maybe you meant a different word there?

king-11 replied:
It should be `tried`

mppf commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1282 @@
+  .. code-block:: Chapel
+
+    const timeout = new timeval(4,0);
+    const connectedClient = connect("google.com", "tcp", IPFamily.IPv4, timeout);
```
TCP is not a service name - maybe the example would be better with `"http"` for example?

mppf commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1282 @@
+  var hints = new sys_addrinfo_t();
+  hints.ai_family = family:c_int;
+  hints.ai_socktype = SOCK_STREAM;
+  var err = getaddrinfo(host.c_str(), service.c_str(), hints, result);
```
the DNS lookup might itself take a long time and I would expect that to be subject to `timeout`. However, I don't know how to implement that. Can libevent do DNS lookup as an event?

king-11 replied:
marked as future work

mppf commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1282 @@
+      c_free(buffer);
+      throw SystemError.fromSyserr(err_out, "recv failed");
+    }
+    return this.recvfrom(bufferLen, timeout, flags);
```
Let's use a loop here, please

mppf commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1282 @@
+      }
+      throw SystemError.fromSyserr(err_out, "send failed");
+    }
+    return this.send(data, address, timeout);
```
Let's write this with a loop please

mppf commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1282 @@
+
+/*
+  Send `data` over socket to the provided address and
+  return number of bytes sent if successful.
```
The way it is written now it looks like it might send only some bytes but not all of the passed bytes.
To send all of the passed bytes I think you would have to have a loop that tries again if only say 1 byte is sent by the sendto call at a time.

king-11 replied:
we can have another function like `sendAll` which could do that because with udp we are working with packets instead of loads of data so if user wants to send large data they can use `sendAll`. ref: This is how python does it

mppf replied:
Right. I'd support a `sendAll` and also a comment here along the lines of this from those Python docs

>  Applications are responsible for checking that all data has been sent; if only some of the data was transmitted, the application needs to attempt delivery of the remaining data. For further information on this topic, consu

mppf commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1282 @@
+
+inline operator !=(in lhs: udpSocket,in rhs: udpSocket) {
+  return lhs.socketFd != rhs.socketFd;
+}
```
Let's add `==`

mppf commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1282 @@
+/*
+  Enables or disables Nagle's algorithm on a given TCP Listener.
+*/
+proc naggle(ref socket: tcpListener, enable:bool = true) throws {
```
I am expecting these to be methods on tcpListener / tcpConn. E.g. here you can write `proc ref tcpListener.nagle(enable: bool = true)`. Arguably in the design https://github.com/chapel-lang/chapel/pull/17899/files#diff-21447e19620654704318fbb1fb753236379e4d6861085d3bb4de8b7a18830199R136 it was unclear if it is a method or not.

Also in the design doc it is `nagle` rather than `naggle`

king-11 replied:
I seemed to have used method in place of procedure but yeah I will update them to be methods

king-11 replied:
And it should be `naggle` i will update it

king-11 commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1282 @@
+/*
+  Overload for :proc:`setSockOpt` that allows setting a :mod:`bytes <Bytes>` value
+  on socket option. It is up to the caller to ensure that the
+  `value` which is a :type::mod:`bytes <Bytes>` parameter contains the proper bits.
```
python has three overloads for setSockOpt so i transferred all three of them

mppf replied:
Sure. It would still be good to show an example that uses it in the docs.

king-11 replied:
In python they have capibility of writing c structs as bytestrings so if we have it in chapel then only this procedure will be useful for us

mppf replied:
Sure. But can you show an example for a setsockopt call that would use such a C structure that's more complex than an int or pointer?

mppf replied:
@king-11 you don't have to be able to show the example in Chapel. I am just interested in an example of when this overload would be used, even if it would be hard or awkward to use in Chapel today. I.e. can you name a setsockopt call that uses such a struct encoded as bytes? That we can learn more about with the C documentation?

king-11 replied:
@mppf as per the `setsockopt` there are three options that need to use C structs:

- **SO_LINGER** uses `linger` struct
- **SO_RCVTIMEO** uses `timeval` struct
- **SO_SNDTIMEO** uses `timeval` struct

mppf replied:
Sure. Can you just mention these specifically in the docs comment? E.g.

```
  Overload for :proc:`setSockOpt` that allows setting a :mod:`bytes <Bytes>` value
  on socket option.
  This is useful for `setsockopt` calls that work with a C struct, including `SO_LINGER`, `SO_RCVTIMEO`, and `SO_SNDTIMEO`.
  It is up to the caller to ensure that the
  `value` which is a :type::mod:`bytes <Bytes>` parameter contains the proper bits.

mppf commented on test/library/standard/Socket/connectblocking.chpl:
```diff
@@ -0,0 +1,148 @@
+
+proc test_connection_ipv4_dns(test: borrowed Test) throws {
+  var port:uint(16) = 8811;
+  var host = "127.0.0.1";
```
Not actually using a DNS name here - do you want to use `localhost` here for this test?

mppf commented on test/library/standard/Socket/connectblocking.chpl:
```diff
@@ -0,0 +1,148 @@
+  var host = "127.0.0.1";
+  var address = ipAddr.ipv4(IPv4Localhost, port);
+  var server = listen(address);
+  begin {
```
I would expect that any time you are using `begin` that you also enclose the code in a `sync` block. For these tests, it would be reasonable to put a `sync` block around the test body. (I don't remember if the unit test infrastructure always adds `sync` blocks, though).

king-11 replied:
For this I think we don't actually need `sync` because the calls made in main thread are blocking and rely on `begin` block so it will not proceed further without completing which might be a good test as well because we don't want to proceed (similar to sync)

mppf replied:
Sure. But my concern is - what if something goes wrong? If we want to check that something in one task happen before something in another task, we can check that (e.g. with an atomic variable). But if something goes wrong and the blocking does not occur, will the test core-dump?

mppf commented:
Relatedly, we talked about having `send` and `recv` for a `tcpConn` aka `file`. Let's add these here so that we can test TCP connections. To get the I/O system properly integrated, we will need to arrange to call these during the other I/O operations (e.g. writeln).

mppf commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1325 @@
+  pragma "no doc"
+  proc init(socketFd: c_int) {
+    this.socketFd = socketFd;
+  }
```
Your tcpListener also needs a `deinit` which probably should close the socket.
It's important to still have a `close` method since `close` can fail (because there is no way of getting the returned error out of a `deinit`).

mppf commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1325 @@
+  Close the file descriptor
+*/
+proc tcpListener.close() {
+  sys_close(this.socketFd);
```
This should be throwing and should consume the return code from sys_close and throw if it was not 0

mppf commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1325 @@
+  proc init(family: IPFamily = IPFamily.IPv4) {
+    this.socketFd = -1;
+    try! this.socketFd = socket(family, SOCK_DGRAM);
+  }
```
As with tcpListener, this record also needs a `deinit` which probably should close the socket.
It's important to still have a `close` method since `close` can fail (because there is no way of getting the returned error out of a `deinit`).

mppf commented:
Related to the above, I'm seeing that test/library/standard/Socket/udpio_ipv4.chpl hangs with `CHPL_TASKS=fifo`, but it works with the default of `CHPL_TASKS=qthreads`. Is this a known issue? It seems to me that it should work with `CHPL_TASKS=fifo` as well.

I also see that the compiler crashes if we try to compile the Socket module with the LLVM backend. That seems like something we need to call out in the PR message.

e-kayrakli commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1403 @@
+  return getsockname(this.socketFd);
+}
+
+inline operator !=(const ref lhs: tcpListener,const ref rhs: tcpListener) {
```
no doc this and the below?

e-kayrakli commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1403 @@
+  return !(lhs.socketFd != rhs.socketFd);
+}
+
+proc tcpListener.writeThis(f) throws {
```
`no doc`

e-kayrakli commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1403 @@
+  It is calulated as min(`SOMAXCONN`, 128) where `SOMAXCONN` is
+  the maximum number of allowed pending connections in the system.
+*/
+var backlogDefault:uint(16) = (if SOMAXCONN <= 128 then SOMAXCONN else 128):uint(16);
```
Is this something that can change? Or be changed by the user?

Consider making it a `const` and/or `no doc`.

e-kayrakli commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1403 @@
+extern const SOMAXCONN: int;
+/*
+  Default `backlog` value used in :proc:`listen`
+  It is calulated as min(`SOMAXCONN`, 128) where `SOMAXCONN` is
```
"calculated"

e-kayrakli commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1403 @@
+  }
+  else {
+    err_out = event_add(writerEvent, c_ptrTo(timeout));
+  }
```
It could help throwing some empty lines between logical parts of the code. But nonetheless, this if/else is just

```chapel
err_out = event_add(writerEvent,
                    if timeout.tv_sec==-1 then nil else c_ptrTo(timeout));
```

e-kayrakli commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1403 @@
+    event_free(internalEvent);
+  }
+
+  while err_out != 0 {
```
same question as above, should we be worried about yielding here?

king-11 replied:
The function call in `readFE()` which yields the thread and that will yield and only return when the running event loop passes down an event to the `sync` variable or a timeout occurs which is also passed down by the event loop to `sync` variable.

e-kayrakli commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1403 @@
+  :rtype: `tcpConn`
+  :throws SystemError: Upon failure to connect.
+*/
+proc connect(const ref address: ipAddr, in timeout = new timeval(-1,0)): tcpConn throws {
```
Should we allow `real`s as timeouts?

king-11 replied:
I think a structure holding these values is a bit more intuitive otherwise both at user and library level we need to do trivial tranformation for getting `seconds` and `microseconds` out of the real value.

e-kayrakli replied:
I didn't do a lot of socket programming but:

> I think a structure holding these values is a bit more intuitive

I disagree with this very strongly :)

mppf replied:
@e-kayrakli - in an offline conversation I had with @king-11 it's my understanding that he's going to look at providing support for both these structures as well as `real` to represent time in seconds.

e-kayrakli replied:
In case I wasn't clear; I have no problem with the current signature, especially if it is following some sort of precedence and/or have other motivations. I was just suggesting of adding a `real` version alongside this one (which what you're saying).

e-kayrakli commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1403 @@
+  :throws SystemError: Upon failure to send any data
+                        within given `timeout`.
+*/
+proc udpSocket.send(data: bytes, in address: ipAddr, in timeout = new timeval(-1,0)):ssize_t throws {
```
This function can really use some vertical spacing.

e-kayrakli commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1403 @@
+    }
+    else {
+      err_out = event_add(internalEvent, c_ptrTo(timeout));
+    }
```
another `if/else` that can be done in a more lightweight manner.

e-kayrakli commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1403 @@
+  return length;
+}
+
+inline operator !=(const ref lhs: udpSocket,const ref rhs: udpSocket) {
```
Similar, `no doc` or document and/or mention the support for comparison in a header comment.

e-kayrakli commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1403 @@
+proc setSockOpt(ref socket: udpSocket, level: c_int, optname: c_int, value: c_int) throws {
+  var socketFd = socket.socketFd;
+  setSockOpt(socketFd, level, optname, value);
+}
```
For all these 3 overloads, I think a better way is to `no doc` the 2nd and 3rd ones, make the 1st one's argument generic, add a `where` clause to it to type-check, change its documentation to say that it accepts 3 different types.

e-kayrakli replied:
I see that this overloads go on and on a bit more. I think we should think about reducing this number. To be clear, I talk about the documentation, I don't mind about the number of overloads in code.

e-kayrakli commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1403 @@
+proc getSockOpt(ref socket: udpSocket, level: c_int, optname: c_int, buflen): bytes throws {
+  var socketFd = socket.socketFd;
+  return getSockOpt(socketFd, level, optname, buflen) ;
+}
```
If we make any changes to `setSockOpt`, we should make symmetric changes to `getSockOpt`

e-kayrakli commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1403 @@
+  :rtype: `ipAddr`
+  :throws SystemError: If socket is not connected
+*/
+proc getpeername(ref socket: tcpConn): ipAddr throws {
```
This might have been discussed elsewhere, but why not `getPeerName`?

e-kayrakli commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1403 @@
+  :rtype: `ipAddr`
+  :throws SystemError: If socket is closed.
+*/
+proc getsockname(ref socket: tcpConn): ipAddr throws {
```
Same. Especially considering that we have `getSockOpt`, `getsockname` feels really wrong to me.

e-kayrakli commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1403 @@
+proc bind(ref socket: tcpConn, ref address: ipAddr, reuseAddr = true) throws {
+  var socketFd = socket.socketFd;
+  bind(socketFd, address, reuseAddr);
+}
```
Same issue: multiple overloads with trivial differences in the documentation.

e-kayrakli commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1403 @@
+proc ref tcpConn.setNagle(enable:bool) throws {
+  var socketFd = this.socketFd;
+  nagle(socketFd, enable);
+}
```
These methods are too far from the types that they are on. Why? Same goes for setDelayAck below.

king-11 replied:
Earlier they were separate function the idea was to keep the documentation about Naggle optimization at one place only and keeping them at separate places will require us to add documentation at all three places(?) or we can reference the documentation for Naggle in other two 🤔

e-kayrakli replied:
Hmm, I don't know anything about this optimization, and it may make sense if there's such a context. I'll leave the call to you and @mppf

mppf replied:
I think it makes sense to have the methods on tcpConn appear in one place in the documentation. We can create a section on Nagle in the module-level docs comment to refer to from these.

e-kayrakli reviewed (commented):
Thanks, @king-11!

My main concern is about how documentation is rendered, and I hope that you can address some of that so that:

- Things are `no doc` ed properly.
- methods on a type are together in documentation
- Ideally, number of overloads in the documentation is reduced.

I have some other inline comments, as well.

king-11 commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1403 @@
+
+  .. code-block:: Chapel
+
+    ipAddr.create("127.0.0.1", 8111, IPFamily.IPv4)
```
Yeah I think this example seems redundant. Will remove it

king-11 commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1403 @@
+  return new ipAddr(addressStorage);
+}
+
+/* compare ipAddr */
```
Going with 1st option and adding details in the doc comment for `ipAddr`

king-11 commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1403 @@
+    event_free(internalEvent);
+  }
+
+  while err_out != 0 {
```
No, the yielding is being handled by call to `localSync$.readFE()`

e-kayrakli replied:
Ah, right.

king-11 commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1403 @@
+  :return: Returns remote address.
+  :rtype: `ipAddr`
+*/
+proc const ref tcpConn.addr throws {
```
this calls in the method on `sockaddr_storage` which is a const ref

e-kayrakli replied:
I am not following.

king-11 replied:
Internally it calls in `numericHost` and `port` which are defined in the `Sys` module and they are declared with const ref so this functions to call those requires const ref as well.

king-11 replied:
Oh alright I got it now this was same need not be const ref normal intent will work fine I was a bit confused with `sockaddr_storage` and this

king-11 commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1403 @@
+private extern proc evthread_use_pthreads();
+
+evthread_use_pthreads();
+var event_loop_base = event_base_new();
```
I will put the code at the end of module definition but a future work for this module includes shifting all the libevent code into a separate module of its own.

king-11 commented on modules/standard/Socket.chpl:
```diff
@@ -0,0 +1,1403 @@
+  :return: Returns file descriptor.
+  :rtype: `int(32)`
+*/
+proc const ref tcpConn.socketFd throws {
```
made the changes.

king-11 commented:
@e-kayrakli @mppf changes were made as per review.

e-kayrakli commented:
I looked at my notes in the PR and the recent commits, and they look good. Thanks @king-11!

king-11 commented:
@mppf yes the tests are compiling and executing I think there can be issues with the `send_http` one in `tcpio.chpl` because it connects to an external server which can reject our requests too maybe so it can be the flaky one but we certainly need it to test the remote connection. So can you try running it again

mppf commented:
@king-11 I don't think we should be connecting to google.com to run these tests. Instead, I think we should run a local Python webserver for the purposes of that test, or maybe remove it entirely. You can see an example of running our own webserver in test/library/packages/Curl/RunServer.chpl .

mppf commented:
I am still seeing timeouts. `tcpio.chpl` worked for me when I tested it on its own but now I am seeing `connectblocking.chpl` time out. Perhaps before I was not using C backend + qthreads.

king-11 commented:
@mppf below are the updates:
- I have updated the module to be a Package instead of Standard Module
- I have added in SKIPIF for whole tests directory by python checking for libevent as a bit hackish not sure if it will work always
- Regarding tests I tired cleaning up my whole repo and making the whole project and running again seems to be running fine 🤔

mppf commented on test/library/standard/Socket/SKIPIF:
```diff
@@ -0,0 +1,8 @@
+import os
+import subprocess
+
+has_libevent = subprocess.getoutput("ldconfig -v | grep libevent - | wc -l")
```
Normally we use pkg-config for such things

mppf replied:
Also ldconfig -v and pkg-config won't exist on all systems, so you have to fail gracefully in that event

mppf replied:
See test/library/packages/Crypto.skipif for an example

mppf commented on test/library/standard/Socket/SKIPIF:
```diff
@@ -0,0 +1,8 @@
+
+has_libevent = subprocess.getoutput("ldconfig -v | grep libevent - | wc -l")
+has_libevent = has_libevent.split("\n")
+print(os.getenv('CHPL_TASKS') != 'fifo' and os.getenv('CHPL_TARGET_COMPILER') != 'llvm' and int(has_libevent[-1]) != 0)
```
We need to skip if CHPL_TASKS != qthreads, right? This script needs to return True if the tests should be skipped and it doesn't look like it's doing that. Maybe it is reversed.

king-11 replied:
Yep made changes

mppf commented:
I tried running on a Mac with Homebrew installed libevent. (To do that I had to change the paths in the require lines in Socket.chpl). I see many failures:

<details>

```
[Executing diff connectblocking.good connectblocking.exec.out.tmp]
2c2
< Flavour: OK
---
> Flavour: ERROR
3a4
> SystemError: Invalid argument (Failed to bind Socket)
6c7
< Flavour: OK
---
> Flavour: ERROR
7a9
> SystemError: Invalid argument (Failed to bind Socket)
10c12
< Flavour: OK
---
> Flavour: FAIL
11a14,22
> AssertionError: Strings differ: 'Invalid argument (connect() failed)' != 'Connection refused (connect() failed)'
> First differing element at index 0:
> 'I'
> 'C'
>
> Second String contains 2 additional elements.
> First extra element is at index 36
> ')'
>
14,24c25,27
< Flavour: OK
< ======================================================================
< ----------------------------------------------------------------------
< test_connection_ipv6_ipaddr()
< Flavour: OK
< ======================================================================
< ----------------------------------------------------------------------
< test_fail_ipv6_noserver()
< Flavour: OK
< ======================================================================
< ----------------------------------------------------------------------
---
> uncaught IllegalArgumentError: illegal argument 'bad cast from int '30' to enum 'IPFamily': '
>   connectblocking.chpl:51: thrown here
>   connectblocking.chpl:51: uncaught here
[Error matching program output for library/packages/Socket/connectblocking]
[Elapsed time to compile and execute all versions of "library/packages/Socket/connectblocking" - 22.780 seconds]
[test: library/packages/Socket/connecttimeout.chpl]
[Executing compiler /Users/mferguson/w/1/bin/darwin-x86_64/chpl -o connecttimeout --cc-warnings connecttimeout.chpl < /dev/null]
[Elapsed compilation time for "library/packages/Socket/connecttimeout" - 21.866 seconds]
[Success compiling library/packages/Socket/connecttimeout]
[Executing program ./connecttimeout  < /dev/null]
[Elapsed execution time for "library/packages/Socket/connecttimeout" - 0.140 seconds]
[Executing diff connecttimeout.good connecttimeout.exec.out.tmp]
2c2
< Flavour: OK
---
> Flavour: ERROR
3a4
> SystemError: Invalid argument (Failed to bind Socket)
6c7
< Flavour: OK
---
> Flavour: ERROR
7a9
> SystemError: Invalid argument (Failed to bind Socket)
10,16c12,14
< Flavour: OK
< ======================================================================
< ----------------------------------------------------------------------
< test_fail_backlog_ipv6()
< Flavour: OK
< ======================================================================
< ----------------------------------------------------------------------
---
> uncaught IllegalArgumentError: illegal argument 'bad cast from int '30' to enum 'IPFamily': '
>   connecttimeout.chpl:42: thrown here
>   connecttimeout.chpl:42: uncaught here
[Error matching program output for library/packages/Socket/connecttimeout]
[Elapsed time to compile and execute all versions of "library/packages/Socket/connecttimeout" - 22.012 seconds]
[test: library/packages/Socket/ipaddr.chpl]
[Executing compiler /Users/mferguson/w/1/bin/darwin-x86_64/chpl -o ipaddr --cc-warnings ipaddr.chpl < /dev/null]
[Elapsed compilation time for "library/packages/Socket/ipaddr" - 22.190 seconds]
[Success compiling library/packages/Socket/ipaddr]
[Executing program ./ipaddr  < /dev/null]
[Elapsed execution time for "library/packages/Socket/ipaddr" - 0.130 seconds]
[Executing diff ipaddr.good ipaddr.exec.out.tmp]
6c6
< Flavour: OK
---
> Flavour: ERROR
7a8
> IllegalArgumentError: Incompatible Address and Family
14c15
< Flavour: OK
---
> Flavour: ERROR
15a17
> IllegalArgumentError: Incompatible Address and Family
22,32c24,26
< Flavour: OK
< ======================================================================
< ----------------------------------------------------------------------
< test_ipaddr_sockaddr_ipv4()
< Flavour: OK
< ======================================================================
< ----------------------------------------------------------------------
< test_ipaddr_sockaddr_ipv6()
< Flavour: OK
< ======================================================================
< ----------------------------------------------------------------------
---
> uncaught IllegalArgumentError: illegal argument 'bad cast from int '30' to enum 'IPFamily': '
>   ipaddr.chpl:60: thrown here
>   ipaddr.chpl:60: uncaught here
[Error matching program output for library/packages/Socket/ipaddr]
[Elapsed time to compile and execute all versions of "library/packages/Socket/ipaddr" - 22.325 seconds]
[test: library/packages/Socket/ipfamily.chpl]
[Executing compiler /Users/mferguson/w/1/bin/darwin-x86_64/chpl -o ipfamily --cc-warnings ipfamily.chpl < /dev/null]
[Elapsed compilation time for "library/packages/Socket/ipfamily" - 22.178 seconds]
[Success compiling library/packages/Socket/ipfamily]
[Executing program ./ipfamily  < /dev/null]
[Elapsed execution time for "library/packages/Socket/ipfamily" - 0.184 seconds]
[Executing diff ipfamily.good ipfamily.exec.out.tmp]
2c2
< Flavour: OK
---
> Flavour: FAIL
3a4
> AssertionError: assert failed - '10' != '30'
[Error matching program output for library/packages/Socket/ipfamily]
[Elapsed time to compile and execute all versions of "library/packages/Socket/ipfamily" - 22.368 seconds]
[test: library/packages/Socket/listen.chpl]
[Executing compiler /Users/mferguson/w/1/bin/darwin-x86_64/chpl -o listen --cc-warnings listen.chpl < /dev/null]
[Elapsed compilation time for "library/packages/Socket/listen" - 22.597 seconds]
[Success compiling library/packages/Socket/listen]
[Executing program ./listen  < /dev/null]
[Elapsed execution time for "library/packages/Socket/listen" - 0.214 seconds]
[Executing diff listen.good listen.exec.out.tmp]
2c2
< Flavour: OK
---
> Flavour: ERROR
3a4
> SystemError: Invalid argument (Failed to bind Socket)
6,8c7,9
< Flavour: OK
< ======================================================================
< ----------------------------------------------------------------------
---
> uncaught IllegalArgumentError: illegal argument 'bad cast from int '30' to enum 'IPFamily': '
>   listen.chpl:20: thrown here
>   listen.chpl:20: uncaught here
[Error matching program output for library/packages/Socket/listen]
[Elapsed time to compile and execute all versions of "library/packages/Socket/listen" - 22.817 seconds]
[test: library/packages/Socket/networkoptimization.chpl]
[Executing compiler /Users/mferguson/w/1/bin/darwin-x86_64/chpl -o networkoptimization --cc-warnings networkoptimization.chpl < /dev/null]
[Elapsed compilation time for "library/packages/Socket/networkoptimization" - 22.353 seconds]
[Executing diff networkoptimization.good networkoptimization.comp.out.tmp]
1,16c1,10
< test_nagle_tcpListener()
< Flavour: OK
< ======================================================================
< ----------------------------------------------------------------------
< test_nagle_tcpConn()
< Flavour: OK
< ======================================================================
< ----------------------------------------------------------------------
< test_delayAck_tcpListener()
< Flavour: OK
< ======================================================================
< ----------------------------------------------------------------------
< test_delayAck_tcpConn()
< Flavour: OK
< ======================================================================
< ----------------------------------------------------------------------
---
> In file included from /var/folders/b6/1lwlz22n5d53878_x52vp3kr0007pg/T//chpl-mferguson.deleteme-6qpTIz/_main.c:60:
> /var/folders/b6/1lwlz22n5d53878_x52vp3kr0007pg/T//chpl-mferguson.deleteme-6qpTIz/Socket.c:1658:46: error: use of undeclared identifier 'TCP_QUICKACK'
> setSockOpt_chpl(socketFd_chpl2, IPPROTO_TCP, TCP_QUICKACK, &c_enable_chpl, &error_chpl, _ln_chpl, _fn_chpl);
>                                              ^
> /var/folders/b6/1lwlz22n5d53878_x52vp3kr0007pg/T//chpl-mferguson.deleteme-6qpTIz/Socket.c:1676:64: error: use of undeclared identifier 'TCP_QUICKACK'
> call_tmp_chpl39 = getSockOpt_chpl(socketFd_chpl2, IPPROTO_TCP, TCP_QUICKACK, &error_chpl, _ln_chpl, _fn_chpl);
>                                                                ^
> 2 errors generated.
> make: *** [/var/folders/b6/1lwlz22n5d53878_x52vp3kr0007pg/T//chpl-mferguson.deleteme-6qpTIz/networkoptimization.tmp] Error 1
> error: compiling generated source
[Error matching compiler output for library/packages/Socket/networkoptimization]
[Elapsed time to compile and execute all versions of "library/packages/Socket/networkoptimization" - 22.358 seconds]
[test: library/packages/Socket/tcpio.chpl]
[Executing compiler /Users/mferguson/w/1/bin/darwin-x86_64/chpl -o tcpio --cc-warnings tcpio.chpl < /dev/null]
[Elapsed compilation time for "library/packages/Socket/tcpio" - 23.848 seconds]
[Success compiling library/packages/Socket/tcpio]
[Executing program ./tcpio  < /dev/null]
[Elapsed execution time for "library/packages/Socket/tcpio" - 0.202 seconds]
[Executing diff tcpio.good tcpio.exec.out.tmp]
2c2
< Flavour: OK
---
> Flavour: ERROR
3a4
> SystemError: Invalid argument (Failed to bind Socket)
6c7
< Flavour: OK
---
> Flavour: ERROR
7a9
> SystemError: Invalid argument (Failed to bind Socket)
10,20c12,14
< Flavour: OK
< ======================================================================
< ----------------------------------------------------------------------
< send_number()
< Flavour: OK
< ======================================================================
< ----------------------------------------------------------------------
< send_http()
< Flavour: OK
< ======================================================================
< ----------------------------------------------------------------------
---
> uncaught IllegalArgumentError: illegal argument 'bad cast from int '30' to enum 'IPFamily': '
>   tcpio.chpl:45: thrown here
>   tcpio.chpl:45: uncaught here
[Error matching program output for library/packages/Socket/tcpio]
[Elapsed time to compile and execute all versions of "library/packages/Socket/tcpio" - 24.055 seconds]
[test: library/packages/Socket/udpio_ipv4.chpl]
[Executing compiler /Users/mferguson/w/1/bin/darwin-x86_64/chpl -o udpio_ipv4 --cc-warnings udpio_ipv4.chpl < /dev/null]
[Elapsed compilation time for "library/packages/Socket/udpio_ipv4" - 23.674 seconds]
[Success compiling library/packages/Socket/udpio_ipv4]
[Executing program ./udpio_ipv4  < /dev/null]
[Elapsed execution time for "library/packages/Socket/udpio_ipv4" - 0.128 seconds]
[Executing diff udpio_ipv4.good udpio_ipv4.exec.out.tmp]
2c2
< Flavour: OK
---
> Flavour: ERROR
3a4
> SystemError: Invalid argument (Failed to bind Socket)
6c7
< Flavour: OK
---
> Flavour: ERROR
7a9
> SystemError: Invalid argument (Failed to bind Socket)
10c12
< Flavour: OK
---
> Flavour: ERROR
11a14
> SystemError: Invalid argument (Failed to bind Socket)
14c17
< Flavour: OK
---
> Flavour: ERROR
15a19
> SystemError: Invalid argument (Failed to bind Socket)
22c26
< Flavour: OK
---
> Flavour: ERROR
23a28
> SystemError: Invalid argument (Failed to bind Socket)
26c31
< Flavour: OK
---
> Flavour: ERROR
27a33
> SystemError: Invalid argument (Failed to bind Socket)
[Error matching program output for library/packages/Socket/udpio_ipv4]
[Elapsed time to compile and execute all versions of "library/packages/Socket/udpio_ipv4" - 23.808 seconds]
[test: library/packages/Socket/udpio_ipv6.chpl]
[Executing compiler /Users/mferguson/w/1/bin/darwin-x86_64/chpl -o udpio_ipv6 --cc-warnings udpio_ipv6.chpl < /dev/null]
[Elapsed compilation time for "library/packages/Socket/udpio_ipv6" - 22.775 seconds]
[Success compiling library/packages/Socket/udpio_ipv6]
[Executing program ./udpio_ipv6  < /dev/null]
[Elapsed execution time for "library/packages/Socket/udpio_ipv6" - 0.138 seconds]
[Executing diff udpio_ipv6.good udpio_ipv6.exec.out.tmp]
2c2
< Flavour: OK
---
> Flavour: ERROR
3a4
> IllegalArgumentError: Incompatible Address and Family
6c7
< Flavour: OK
---
> Flavour: ERROR
7a9
> IllegalArgumentError: Incompatible Address and Family
10c12
< Flavour: OK
---
> Flavour: ERROR
11a14
> IllegalArgumentError: Incompatible Address and Family
14c17
< Flavour: OK
---
> Flavour: ERROR
15a19
> IllegalArgumentError: Incompatible Address and Family
18c22
< Flavour: OK
---
> Flavour: ERROR
19a24
> IllegalArgumentError: Incompatible Address and Family
22c27
< Flavour: OK
---
> Flavour: ERROR
23a29
> IllegalArgumentError: Incompatible Address and Family
26c32
< Flavour: OK
---
> Flavour: ERROR
27a34
> IllegalArgumentError: Incompatible Address and Family
[Error matching program output for library/packages/Socket/udpio_ipv6]
[Elapsed time to compile and execute all versions of "library/packages/Socket/udpio_ipv6" - 22.918 seconds]
[Finished subtest "library/packages/Socket" - 206.686 seconds]

[Done with tests - 211114.134852]
[Log file: /Users/mferguson/w/1/test/Logs/mferguson.darwin.log ]

[Test Summary - 211114.134852]
[Error matching program output for library/packages/Socket/connectblocking]
[Error matching program output for library/packages/Socket/connecttimeout]
[Error matching program output for library/packages/Socket/ipaddr]
[Error matching program output for library/packages/Socket/ipfamily]
[Error matching program output for library/packages/Socket/listen]
[Error matching compiler output for library/packages/Socket/networkoptimization]
[Error matching program output for library/packages/Socket/tcpio]
[Error matching program output for library/packages/Socket/udpio_ipv4]
[Error matching program output for library/packages/Socket/udpio_ipv6]
[Warning: SKIPIF error.]
```

</details>

The testing I previously told you about failing connectblocking.chpl is using Ubuntu 21.10. It still seems to be timing out.

Can you use some VMs or Docker images or something to try it on some different systems? While I believe it is working on your system it seems it does not work on other systems. (In fact I have of 3 systems I've tried it on, I've seen problems in all 3).

king-11 commented:
- I tried to run the tests on a new Ubuntu 20.04 VM seems to except that `tcpio.chpl` is hanging on the VM individually and with all the tests as well.
- The macOS issue I think the numbers that I have hardcoded in IPFamily are responsible

mppf commented:
thanks @king-11. I think it's important to get all of the tests passing somewhere reproduce-able before we merge it. Can you look into the tcpio.chpl hang you are seeing?

king-11 commented on test/library/packages/Socket/tcpio.chpl:
```diff
@@ -0,0 +1,96 @@
+  }
+}
+
+proc send_number(test: borrowed Test) throws {
+  var port:uint(16) = 6000;
+  var address = ipAddr.ipv4(IPv4Localhost, port);
+  var server = listen(address);
+  sync {
+    begin {
+      var conn = server.accept();
+      var reader = conn.reader();
+      var y:int = reader.read(int);
+      test.assertEqual(y, 42);
+    }
+
+    var conn = connect(address);
+    var writer = conn.writer();
+    writer.write(42, " ");
+  }
+}
```
@mppf I made a local dev env with Ubuntu and tried to work out the issue that I observed in VM. 

It turns this is a pretty flaky error and is not caused by `sent_http` it's by `send_number`. This was something I observed earlier too but I thought was resolved at that time so if you run the test 4/5 times it might run perfectly through all or will be hangup randomly at this test.

The thing that is intriguing is if I remove the `recv_number` test everything works fine without hanging even if you run it 1k times.

mppf commented:
I've run testing on 2 systems with libevent 2.1.12 and all tests are passing and the skipifs seem to be working as expected. Great! I'm not going to merge today though because I'll be away the rest of the week. We'll consider including it in the 1.25.1 point release.
